doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+# RFC 001: Extend Quint with Row-Polymorphic Sum Types
+
+**Table of Contents**
+
+- [tl;dr](#tldr)
+- [Motivation](#motivation)
+- [Context](#context)
+- [Statics](#statics-1)
+- [Dynamics](#dynamics-1)
+- [Concrete Syntax](#concrete-syntax)
+- [Additional consideration](#additional-consideration)
+
+## tl;dr
+
+This section gives a concise overview of the principle proposal. In depth
+discussion follows in the subsequent sections.
+
+### Concrete syntax
+
+Here I illustrate two attractive alternatives for each critical construct. See
+[Syntax](#syntax) for a discussion of motivations and trade offs.
+
+#### Declaration
+
+This group of alternatives is closer to Rust, but with significant variation for
+reasons discussed in [Syntax](#syntax).
+
+```quint
+type T =
+  | A(int)
+  | B(str)
```

Comment:
Just looking at [Rust enum types](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html), I can actually see how they understand sum types to be generalization of C enums. As far as I remember, Haskell also has special keyword for sum types, like `data` or something similar.

If we replace `|` in `match` (see my comment below) with `,`, then the Rust syntax makes sense to me:

```rust
enum T {
  A(int),
  B(str),
}
```

I have noticed it before, but this syntax is actually closer to the variant types in Apalache than I thought.

Reply:
`data` in Haskell differentiates from `type`, for type aliases, and `newtype` a wrapper thing needed cause of their typeclass system. Every user defined type in Haskell is wrapped in (at least) a unary sum type label, so in that case it is kind of related to sum types. But it is clear from ML that there is no need to have different keywords for records and sum types.

I think the syntax in rust just follows that of C++ https://en.cppreference.com/w/cpp/language/enum

But, critically, this is consistent syntax in rust. They have:

```
struct Foo { ... }
mod foo { ... }
enum Foo { ... }
trait Foo { ... }
impl Foo { ... }
```

If we opt for `enum T { ... }` and don't also change our syntax for declaring record types, we lose the consistency with type defs for our records. So I'd just advocate that we be consistent internally.